### ISSUE TITLE
False positive for Optional with if branches

### ISSUE THREAD

**Issue by dedalusj** [python/mypy#5902]
Bug report

* Repro case

```python
def foo(a: Dict, b: Optional[str] = None):
    if b is None and 'k' not in a:
        b = '1'
    elif 'k' in a:
        b = '2'
    c: str = b  # Incompatible types error reported here
    print(c)
```

* What is the actual behavior/output?

Incompatible types error reported on `c` var assignment

* What is the behavior/output you expect?

No error reported

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

`mypy 0.641`. The is present in master as well.

* What are the mypy flags you are using? (For example --strict-optional)

None

**Comment by gvanrossum** [comment]
Looks like you're being too clever here. Mypy doesn't understand that the `elif` clause rules out `b` being `None`. (Honestly I had to do a double take and re-read the code too before I trusted this was correct.)

We don't have plans to add such logic to mypy, so I recommend rewriting your code to be a bit simpler.
